### ISSUE TITLE
Missing dependencies unless `--shamefully-hoist` option is used

### ISSUE THREAD

**Issue by garysassano** [pnpm/pnpm#7615]
### Verify latest release

- [x] I verified that the issue exists in the latest pnpm release

### pnpm version

8.15.1

### Which area(s) of pnpm are affected? (leave empty if unsure)

Dependencies resolver

### package.json
```json
{
  "name": "test",
  "version": "1.0.0",
  "private": true,
  "scripts": {
    "dev": "nuxt",
    "build": "nuxt build",
    "start": "nuxt start",
    "generate": "nuxt generate"
  },
  "dependencies": {
    "core-js": "^3.19.3",
    "express": "^4.17.3",
    "nuxt": "^2.15.8",
    "vue": "^2.6.14",
    "vue-server-renderer": "^2.6.14",
    "vue-template-compiler": "^2.6.14",
    "webpack": "^4.46.0"
  },
  "devDependencies": {
    "@nuxt/types": "^2.15.8",
    "@nuxt/typescript-build": "^2.1.0"
  }
}
```

### Describe the Bug

When using `pnpm i`, all modules' dependants are ignored unless i provide the `--shamefully-hoist` option.

### Expected Behavior

I expected all dependencies to be resolved correctly by default.

### Which Node.js version are you using?

20.10.0

### Which operating systems have you used?

- [ ] macOS
- [ ] Windows
- [ ] Linux

### If your OS is a Linux based, which one it is? (Include the version if relevant)

Ubuntu 22.04.3 LTS on WSL2

**Comment by nachoaldamav** [comment]
The non-direct dependencies are installed too, but only the direct dependencies are available in the public node_modules folder, the indirect dependencies are saved at `node_modules/.pnpm`, this is intended, you can find more information here https://pnpm.io/blog/2020/05/27/flat-node-modules-is-not-the-only-way

**Comment by nachoaldamav** [comment]
You should install the dependencies inside the Dockerfile.

**Comment by garysassano** [comment]
I think the problem I experienced is actually [this](https://github.com/nuxt/nuxt/issues/14146), which got fixed in Nuxt v3.0 but would still be present in lower versions.

**Comment by hifron** [comment]
> I think the problem I experienced is actually [this](https://github.com/nuxt/nuxt/issues/14146), which got fixed in Nuxt v3.0 but would still be present in lower versions.

If it is fixed and nowhere else needed, then it is obsolete. And if it is needed for Electron, it is bug if simply node_modules is not sufficient for **build**.

[Electron exposure](https://github.com/electron-vite/electron-vite-vue/issues/522) and [Error](https://github.com/electron-userland/electron-builder/issues/7702)

also
> The non-direct dependencies are installed too, but only the direct dependencies are available in the public node_modules folder, the indirect dependencies are saved at `node_modules/.pnpm`, this is intended, you can find more information here https://pnpm.io/blog/2020/05/27/flat-node-modules-is-not-the-only-way

and where is link from such usages to correct destination(**in build**)? .pnpm is 3rd party destination without symlinks(and symlinks are not present on all systems...)

_if speed is in question - that's also special question itself..._
